src/unix/coredump.c

```diff
@@ -226,7 +226,7 @@ void coredump(thread t, struct siginfo *si, status_handler complete)
     status s = STATUS_OK;
 
     /* stop running any other user threads */
-    shutting_down = true;
+    shutting_down |= SHUTDOWN_TERM;
     wakeup_or_interrupt_cpu_all();
```

Comment:
These lines should be removed, similarly to what is done in exit_group(). But in order for the core dump to be executed after ensuring that user threads are not running, the core dump needs to be executed as a shutdown completion. So default_signal_action(), instead of calling directly coredump(), should instantiate a shutdown_handler (which will call coredump()) and call add_shutdown_completion(), without calling runloop() afterwards; so these lines:
```
        coredump(t, &qs->si, closure(h, after_dump_halt, t, signum));
        clear_fault_handler();
        runloop();
        return;
```
will become something like:
```
        add_shutdown_completion(closure(h, coredump_handler, t, &qs->si));
        clear_fault_handler();
        break;
```

Reply:
this is the last part yet to be implemented

Reply:
would smth like this be correct
```diff
@@ -1036,6 +1036,18 @@ closure_function(2, 1, void, after_dump_halt,
     closure_finish();
 }
 
+closure_function(4, 2, void, coredump_shutdown_handler,
+                 heap, h, thread, t, struct siginfo *, si, int, signum,
+                 int, status, merge, m)
+{
+    heap h = bound(h);
+    thread t = bound(t);
+    struct siginfo *si = bound(si);
+    int signum = bound(signum);
+
+    coredump(t, si, closure(h, after_dump_halt, t, signum));
+}
+
 static void default_signal_action(thread t, queued_signal qs)
 {
     heap h = heap_locked(get_kernel_heaps());
@@ -1070,10 +1082,9 @@ static void default_signal_action(thread t, queued_signal qs)
     case SIGTRAP:
     case SIGXCPU:
     case SIGXFSZ:
-        coredump(t, &qs->si, closure(h, after_dump_halt, t, signum));
+        add_shutdown_completion(closure(h, coredump_shutdown_handler, h, t, &qs->si, signum));
         clear_fault_handler();
-        runloop();
-        return;
+        break;
 
     case SIGSTOP:
     case SIGTSTP:
```

Reply:
do we need `clear_fault_handler();` there still?